### PR TITLE
Add service_answers support to interaction upload tool

### PR DESCRIPTION
**Pull request by elcct**
### Description of change

This adds support of `service_answers` field for interaction upload tool.

At the moment we only have one question at most that requires only one answer, so that this can be provided with a `service_answer` column in the spreadsheet.

If we get to have more questions, we could extend this to have `service_answer_1`, `service_answer_2` and so on. 

If the service answer is not found, we should get an error `A service answer could not be found with the specified name and service.`.

Service answer is only required when a service that has questions is provided in the CSV file.

Provided answer once found valid it is then transformed into a dictionary like below:

```
{
  <question_id_str>: {
    <answer_id_str>: {}
  }
}
```

the above dictionary is then stored in the Interaction's JSONField `service_answers`.

### Checklist

* [x] Has a new newsfragment been created? Check [changelog/README.rst](https://github.com/uktrade/data-hub-leeloo/blob/master/changelog/README.rst) for instructions
* [ ] Do any added or updated endpoints appear in the API documentation? See [docs/Maintaining the API documentation.md](https://github.com/uktrade/data-hub-leeloo/blob/develop/docs/Maintaining&#32;the&#32;API&#32;documentation.md) for more details
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [ ] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [x] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

File: datahub/interaction/admin_csv_import/row_form.py [hunk 1]
```diff
@@ -218,7 +229,6 @@ def clean(self):
 
         self._validate_not_duplicate_of_prior_row(data)
         self._validate_not_duplicate_of_existing_interaction(data)
-
```
Comment (reupen) [inline datahub/interaction/admin_csv_import/row_form.py]: Was this change intended?

File: datahub/interaction/admin_csv_import/row_form.py [hunk 2]
```diff
@@ -296,6 +306,32 @@ def _populate_adviser(self, data, adviser_field, team_field):
         except ValidationError as exc:
             self.add_error(adviser_field, exc)
 
+    def _populate_service_answers(self, data):
+        """Transform service_answer into service_answers dictionary."""
+        service = data.get('service')
+        if service and service.interaction_questions.count() > 0:
```
Comment (reupen) [inline datahub/interaction/admin_csv_import/row_form.py]: Shouldn't we also return an error if a `service_answer` is provided but the service has no service answers?

File: datahub/interaction/admin_csv_import/row_form.py [hunk 3]
```diff
@@ -370,12 +406,17 @@ def cleaned_data_as_serializer_dict(self):
...
         }
+        if data.get('service_answers'):
+            creation_data['service_answers'] = data['service_answers']
 
         if data['kind'] == Interaction.KINDS.service_delivery:
             creation_data['is_event'] = bool(data.get('event_id'))
 
         return creation_data
 
+    def _validate_service_answer(self, data, service_answer_field):
```
Comment (reupen) [inline datahub/interaction/admin_csv_import/row_form.py]: Is this intended to be here?

File: datahub/interaction/test/admin_csv_import/test_row_form.py [hunk 4]
```diff
@@ -323,6 +325,20 @@ class TestInteractionCSVRowFormValidation:
                 },
                 id='cannot specify event_id for an interaction',
             ),
+            pytest.param(
+                {
+                    'kind': Interaction.KINDS.interaction,
+                    'service':
```
Comment (reupen) [inline datahub/interaction/test/admin_csv_import/test_row_form.py]: I think it would be better to avoid having a dependency on the service and service answer fixtures in these tests. 

It would be good to follow the same pattern as the other tests using factories – there is a similar dependency between the `adviser_1` and `team_1` fields so it should be possible to do it in a similar way.

File: datahub/interaction/admin_csv_import/row_form.py (hunk 1 above)
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: Oh no, reverted. Thanks!

File: datahub/interaction/admin_csv_import/row_form.py (hunk 3 above)
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: I forgot to delete that one, thank you!

File: datahub/interaction/admin_csv_import/row_form.py (hunk 2 above)
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: Good idea - updated. Thanks!

File: datahub/interaction/test/admin_csv_import/test_row_form.py (hunk 4 above)
Comment (elcct) [reply datahub/interaction/test/admin_csv_import/test_row_form.py]: I agree. I have removed the dependency now. Thank you!

File: datahub/interaction/admin_csv_import/row_form.py [hunk 5]
```diff
@@ -296,6 +310,41 @@ def _populate_adviser(self, data, adviser_field, team_field):
...
+        service_answer = data.get('service_answer')
+
+        if service and service.interaction_questions.count() > 0:
+
+            try:
+                service_answer_option_db = ServiceAnswerOption.objects.get(
+                    name=service_answer,
+                    question__service__pk=service.pk,
+                )
+                data['service_answers'] = {
```
Comment (currycoder) [inline datahub/interaction/admin_csv_import/row_form.py]: Could it be good to call this something distinct rather than overwriting the existing `data['service_answers']`?

File: datahub/interaction/admin_csv_import/row_form.py [hunk 6]
```diff
@@ -296,6 +310,41 @@ def _populate_adviser(self, data, adviser_field, team_field):
...
             self.add_error(adviser_field, exc)
 
+    def _populate_service_answers(self, data):
+        """Transform service_answer into service_answers dictionary."""
+        service = data.get('service')
+        service_answer = data.get('service_answer')
+
+        if service and service.interaction_questions.count() > 0:
+
+            try:
```
Comment (currycoder) [inline datahub/interaction/admin_csv_import/row_form.py]: It might be helpful for readability to break the body of this `if` out in to another helper method.  There's a fair amount of indentation.

File: datahub/interaction/admin_csv_import/row_form.py (hunk 5 above)
Comment (currycoder) [reply datahub/interaction/admin_csv_import/row_form.py]: Oh wait this is a new key!  Is it plural because you are anticipating multiple answers in the future?

File: datahub/interaction/admin_csv_import/row_form.py [hunk 7]
```diff
@@ -296,6 +310,41 @@ def _populate_adviser(self, data, adviser_field, team_field):
...
+
+            except ServiceAnswerOption.DoesNotExist:
+                self.add_error(
+                    'service_answer',
+                    ValidationError(
+                        SERVICE_ANSWER_NOT_FOUND,
+                        code='service_answer_not_found',
+                    ),
+                )
+        elif service_answer:
```
Comment (currycoder) [inline datahub/interaction/admin_csv_import/row_form.py]: Might be nice to have a comment here like "We have a service answer, but there was no service or the service did not require an answer"

File: datahub/interaction/admin_csv_import/row_form.py [hunk 8]
```diff
@@ -296,6 +310,41 @@ def _populate_adviser(self, data, adviser_field, team_field):
...
+                    question__service__pk=service.pk,
+                )
+                data['service_answers'] = {
+                    str(service_answer_option_db.question.pk): {
+                        str(service_answer_option_db.pk): {},
+                    },
+                }
+
+            except ServiceAnswerOption.DoesNotExist:
+                self.add_error(
```
Comment (currycoder) [inline datahub/interaction/admin_csv_import/row_form.py]: My understanding is that this will also cover the case where no service answer was given, but it was required for the service.  It might be good to call that situation out with a separate error message.

File: datahub/interaction/admin_csv_import/row_form.py (hunk 5 above)
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: It is plural on the Interaction model, because it can hold multiple questions and answers and in the CSV file we support only one, so the name is singular.

File: datahub/interaction/test/admin_csv_import/test_row_form.py [hunk 9]
```diff
@@ -1326,6 +1385,56 @@ def test_save_rolls_back_on_error(self, monkeypatch):
...
+            'service_answer': service_answer.name,
+            'communication_channel': communication_channel.name,
+        }
+        form = InteractionCSVRowForm(data=data)
+
+        assert form.is_valid()
+        interaction = form.save(user, source=source)
+        interaction.refresh_from_db()
+
+        assert interaction.theme == data['theme']
```
Comment (reupen) [inline datahub/interaction/test/admin_csv_import/test_row_form.py]: We could probably make this a little bit more specific and just check the relevant service answer attributes.

Also would be good to move the test up a tiny bit (near the other similar tests).

File: datahub/interaction/admin_csv_import/row_form.py (hunk 6 above)
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: I have reorganised the logic so now it should be more readable now.

File: datahub/interaction/admin_csv_import/row_form.py (hunk 7 above)
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: I have reorganised it a bit so maybe now the comment will not be needed?

File: datahub/interaction/admin_csv_import/row_form.py (hunk 8 above)
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: I have added this case. Thanks!

File: datahub/interaction/admin_csv_import/row_form.py [hunk 10]
```diff
@@ -296,6 +310,41 @@ def _populate_adviser(self, data, adviser_field, team_field):
         except ValidationError as exc:
             self.add_error(adviser_field, exc)
 
+    def _populate_service_answers(self, data):
+        """Transform service_answer into service_answers dictionary."""
+        service = data.get('service')
+        service_answer = data.get('service_answer')
+
+        if service and service.interaction_questions.count() > 0:
```
[the remark this answers is not in the document]
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: This is a good idea so I reorganised this method a bit using your suggestions. Thank you!

File: datahub/interaction/admin_csv_import/row_form.py [hunk 11]
```diff
@@ -296,6 +310,41 @@ def _populate_adviser(self, data, adviser_field, team_field):
...
+
+            try:
+                service_answer_option_db = ServiceAnswerOption.objects.get(
+                    name=service_answer,
+                    question__service__pk=service.pk,
```
[the remark this answers is not in the document]
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: Good point! Fixed.

File: datahub/interaction/admin_csv_import/row_form.py [hunk 12]
```diff
@@ -370,6 +419,8 @@ def cleaned_data_as_serializer_dict(self):
             'theme': data['theme'],
             'was_policy_feedback_provided': False,
         }
+        if data.get('service_answers'):
+            creation_data['service_answers'] = data['service_answers']
```
[the remark this answers is not in the document]
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: I think it was breaking few tests if without the `if` and I didn't want to change those at the time, then I forgot to remove it. It is now updated. Thank you!

File: datahub/interaction/test/admin_csv_import/test_row_form.py (hunk 9 above)
Comment (elcct) [reply datahub/interaction/test/admin_csv_import/test_row_form.py]: That's a good suggestion. I moved the test a little and removed irrelevant checks. Thank you!

File: datahub/interaction/admin_csv_import/row_form.py [hunk 13]
```diff
@@ -296,6 +313,54 @@ def _populate_adviser(self, data, adviser_field, team_field):
...
+    def _populate_service_answers(self, data):
+        """Transform service_answer into service_answers dictionary."""
+        service = data.get('service')
+        if not service:
+            return
+
+        service_answer = data.get('service_answer')
+
+        if service.interaction_questions.count() == 0:
+            if service_answer:
```
Comment (reupen) [inline datahub/interaction/admin_csv_import/row_form.py]: Would be good to combine these two if statements e.g. `if service_answer and not service.interaction_questions.exists()`.

File: datahub/interaction/admin_csv_import/row_form.py [hunk 14]
```diff
@@ -296,6 +313,54 @@ def _populate_adviser(self, data, adviser_field, team_field):
...
+                ValidationError(
+                    SERVICE_ANSWER_REQUIRED,
+                    code='service_answer_required',
+                ),
+            )
+            return
+
+        try:
+            service_answer_option_db = ServiceAnswerOption.objects.get(
+                name=service_answer,
```
Comment (reupen) [inline datahub/interaction/admin_csv_import/row_form.py]: Would be good to use `name__iexact` to make it case-insensitive (in line with most of the other fields). (Plus add tests for case sensitivity...)

File: datahub/interaction/test/admin_csv_import/test_row_form.py [hunk 15]
```diff
@@ -1016,6 +1058,36 @@ def test_contact_lookup(self, input_email, matching_status, match_on_alternative
         else:
             assert not contact
 
+    @pytest.mark.parametrize(
+        'service_answer',
```
Comment (reupen) [inline datahub/interaction/test/admin_csv_import/test_row_form.py]: Is this parameter used?

File: datahub/interaction/admin_csv_import/row_form.py (hunk 14 above)
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: Good point. I changed it so now it is not case sensitive and I also adjusted the test.

File: datahub/interaction/test/admin_csv_import/test_row_form.py (hunk 15 above)
Comment (elcct) [reply datahub/interaction/test/admin_csv_import/test_row_form.py]: It wasn't, but now it is being used for checking case sensitivity.

File: datahub/interaction/admin_csv_import/row_form.py (hunk 13 above)
Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: Maybe the indent in the return is not too visible, but this is testing two things - if there is no questions, then we should exit early, but in case there is actually an answer provided, that is an error. If I combine the two, I would have to check again if there is no questions and then exit?

Comment (reupen) [reply datahub/interaction/admin_csv_import/row_form.py]: Oh, yes, sorry, my mistake. Maybe it could deal with the return before adding the error to make it a bit more obvious?

File: datahub/interaction/test/admin_csv_import/test_row_form.py (hunk 15 above)
Comment (reupen) [reply datahub/interaction/test/admin_csv_import/test_row_form.py]: It would be good to not use `service_answer` twice in the function – maybe the argument could be renamed `service_answer_name`.

File: datahub/interaction/admin_csv_import/row_form.py (hunk 13 above)
Comment (reupen) [reply datahub/interaction/admin_csv_import/row_form.py]: Although I guess you'd have two returns then. Maybe just a comment before the return...

Comment (elcct) [reply datahub/interaction/admin_csv_import/row_form.py]: I added a comment. Thank you!

File: datahub/interaction/test/admin_csv_import/test_row_form.py (hunk 15 above)
Comment (elcct) [reply datahub/interaction/test/admin_csv_import/test_row_form.py]: Yes! Nice one. I have renamed it now. Thank you.
